compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,418 @@
+                    }
+                    ty::ExistentialPredicate::Projection(_) => {}
+                    ty::ExistentialPredicate::AutoTrait(_) => {}
+                };
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Placeholder types
+        ty::Bound(..) | ty::Error(_) | ty::Infer(..) | ty::Param(_) | ty::Placeholder(_) => {
+            // u1p as vendor extended type
+            let mut s = String::from("u1p");
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::GeneratorWitness(..) => {
+            bug!("encode_ty: unexpected `GeneratorWitness`");
+        }
+
+        ty::Opaque(..) => {
+            bug!("encode_ty: unexpected `Opaque`");
+        }
+
+        ty::Projection(..) => {
+            bug!("encode_ty: unexpected `Projection`");
+        }
+    };
+
+    Some(typeid.to_string())
+}
+
+/// Returns a type metadata identifier for the specified FnAbi using the Itanium C++ ABI with vendor
+/// extended type qualifiers and types for Rust types that are not used at the FFI boundary.
+pub fn typeid_for_fnabi<'tcx>(tcx: TyCtxt<'tcx>, fn_abi: &FnAbi<'tcx, Ty<'tcx>>) -> Option<String> {
+    // A name is mangled by prefixing "_Z" to an encoding of its name, and in the case of functions
+    // its type.
+    let mut typeid = String::from("_Z");
+
+    // Clang uses the Itanium C++ ABI's virtual tables and RTTI typeinfo structure name as type
+    // metadata identifiers for function pointers. The typeinfo name encoding is a two-character
+    // code (i.e., 'TS') prefixed to the type encoding for the function.
+    typeid.push_str("TS");
+
+    // Function types are delimited by an "F..E" pair
+    typeid.push('F');
+
+    // A dictionary of substitution candidates used for compression (see
+    // https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-compression).
+    let mut dict: FxHashMap<(Ty<'tcx>, TyQ), usize> = FxHashMap::default();
+
+    // Identify C char and integer type uses and encode them with their respective builtin type
+    // encoding as specified by the Itanium C++ ABI for extern function types with the "C" calling
+    // convention to use this encoding for cross-language LLVM CFI.
+    let encode_c_types = match fn_abi.conv {
+        // FIXME(rcvalle): See comment before the ty::Adt match in encode_ty().
+        Conv::C => false,
+        _ => false,
+    };
+
+    // Encode the return type
+    typeid.push_str(&encode_ty(tcx, fn_abi.ret.layout.ty, &mut dict, encode_c_types)?);
+
+    if !fn_abi.args.is_empty() {
+        // Encode the parameter types
+        for arg in fn_abi.args.iter() {
+            let s = encode_ty(tcx, arg.layout.ty, &mut dict, encode_c_types)?;
+            if !s.is_empty() {
+                typeid.push_str(&s);
+            }
+        }
+
+        if fn_abi.c_variadic {
```

Comment:
`fn_abi.args` contains all variadic arguments too at the caller side. It doesn't at the callee side. I assume you need to skip them on the caller side too?

Reply:
If the variadic arguments are in `fn_abi.args` we definitely want to exclude them from the encoding--they're encoded as just 'z'. How can I distinguish them from the non-variadic arguments?

Reply:
The first `fn_abi.fixed_count` arguments are non-variadic I believe.

Reply:
Done.